README.md
invoke-fraction
===============

Partition invocations across a set of functions based on defined fractions.

### Example

```js
// Randomly invoke one of following arrow functions based on the given fraction.
// e.g. it'll invoke the first function ~10% of the time if called repeatedly.
invokeFraction([
  [0.1, () => console.log('10%')],
  [0.3, () => console.log('30%')],
  // The number here doesn't matter, provided it's >0, as it's the last partition.
  [0.6, () => console.log('60%')],
]);
```

The `invokeFraction` function also supports a `defaultHandler` parameter, which stands in for the
last item of the handlers array. Note that this is the recommended way to avoid letting the last
item in the handlers array be invoked for the remaining partition, rather than the defined fraction.

```js
invokeFraction([
  [0.1, () => console.log('10%')],
  [0.3, () => console.log('30%')],
], () => console.log('60%'));
```

The `invokeFraction` function will return the value returned by the function invoked, so you can use
this with promises and async/await.


CHANGELOG.md
## Release History

* 2.0.0 Migrating module to Typescript


test/invoke-fraction.test.js
const invokeFraction = require('../src');

const tosser = (err = new Error('should not be called')) => () => {
  throw err;
};

describe('invokeFraction', () => {
  it('should invoke the default handler', () => {
    const handler = jest.fn().mockReturnValueOnce('value');
    expect(invokeFraction([], handler)).toBe('value');
    expect(handler.mock.calls).toEqual([[]]);
  });

  it('should invoke the last handler', () => {
    const handler = jest.fn().mockReturnValueOnce('value');
    expect(invokeFraction([[1e-10, handler]])).toBe('value');
    expect(handler.mock.calls).toEqual([[]]);
  });

  it('should prioritize the default handler over the last handler', () => {
    const lastHandler = jest.fn().mockReturnValueOnce('bad');
    const defaultHandler = jest.fn().mockReturnValueOnce('good');
    expect(invokeFraction([[1e-10, lastHandler]], defaultHandler)).toBe('good');
    expect(lastHandler.mock.calls).toEqual([]);
    expect(defaultHandler.mock.calls).toEqual([[]]);
  });

  it('should invoke an enabled handler', () => {
    const disabledHandler = jest.fn().mockImplementation(tosser());
    const enabledHandler = jest
      .fn()
      .mockReturnValueOnce('good1')
      .mockReturnValueOnce('good2')
      .mockReturnValueOnce('good3');
    expect(
      invokeFraction([
        [0, disabledHandler],
        [0, disabledHandler],
        [1, enabledHandler],
        [0, disabledHandler],
      ])
    ).toBe('good1');
    expect(
      invokeFraction([
        [0, disabledHandler],
        [0, disabledHandler],
        [1e-10, enabledHandler],
        [0, disabledHandler],
      ])
    ).toBe('good2');
    expect(
      invokeFraction([
        [0, disabledHandler],
        [0, disabledHandler],
        [1e-10, enabledHandler],
        [0, disabledHandler],
        [1, enabledHandler],
      ])
    ).toBe('good3');

    expect(enabledHandler.mock.calls).toEqual([[], [], []]);
  });
});


test/invoke-fraction.test.ts
import invokeFraction from '../src';

const tosser =
  (err = new Error('should not be called')) =>
  () => {
    throw err;
  };

describe('invokeFraction', () => {
  it('should invoke the default handler', () => {
    const handler = jest.fn().mockReturnValueOnce('value');
    expect(invokeFraction([], handler)).toBe('value');
    expect(handler.mock.calls).toEqual([[]]);
  });

  it('should invoke the last handler', () => {
    const handler = jest.fn().mockReturnValueOnce('value');
    expect(invokeFraction([[1e-10, handler]])).toBe('value');
    expect(handler.mock.calls).toEqual([[]]);
  });

  it('should prioritize the default handler over the last handler', () => {
    const lastHandler = jest.fn().mockReturnValueOnce('bad');
    const defaultHandler = jest.fn().mockReturnValueOnce('good');
    expect(invokeFraction([[1e-10, lastHandler]], defaultHandler)).toBe('good');
    expect(lastHandler.mock.calls).toEqual([]);
    expect(defaultHandler.mock.calls).toEqual([[]]);
  });

  it('should invoke an enabled handler', () => {
    const disabledHandler = jest.fn().mockImplementation(tosser());
    const enabledHandler = jest
      .fn()
      .mockReturnValueOnce('good1')
      .mockReturnValueOnce('good2')
      .mockReturnValueOnce('good3');
    expect(
      invokeFraction([
        [0, disabledHandler],
        [0, disabledHandler],
        [1, enabledHandler],
        [0, disabledHandler],
      ])
    ).toBe('good1');
    expect(
      invokeFraction([
        [0, disabledHandler],
        [0, disabledHandler],
        [1e-10, enabledHandler],
        [0, disabledHandler],
      ])
    ).toBe('good2');
    expect(
      invokeFraction([
        [0, disabledHandler],
        [0, disabledHandler],
        [1e-10, enabledHandler],
        [0, disabledHandler],
        [1, enabledHandler],
      ])
    ).toBe('good3');

    expect(enabledHandler.mock.calls).toEqual([[], [], []]);
  });
});


src/index.js
// @flow

type HandlerFunction<T> = () => T;
type Handler<T> = [number, HandlerFunction<T>];
type Handlers<T> = Array<Handler<T>>;

/**
 * Randomly invokes a handler based on the provided fractional call rates. Has a much cleaner
 * recursive implementation in a purely functional language where we'd recurse through a linked list
 * ;).
 *
 * Example:
 *
 * ```js
 * // Randomly invoke one of following arrow functions based on the given fraction.
 * // e.g. it'll invoke the first function ~10% of the time if called repeatedly.
 * invokeFraction([
 *   [0.1, () => console.log('10%')],
 *   [0.3, () => console.log('30%')],
 *   // The number here doesn't matter, provided it's >0, as it's the last partition.
 *   [0.6, () => console.log('60%')],
 * ]);
 * ```
 *
 * The `invokeFraction` function also supports a `defaultHandler` parameter, which stands in for the
 * last item of the handlers array. Note that this is the recommended way to avoid letting the last
 * item in the handlers array be invoked for the remaining partition, rather than the defined
 * fraction.
 *
 * ```js
 * invokeFraction([
 *   [0.1, () => console.log('10%')],
 *   [0.3, () => console.log('30%')],
 * ], () => console.log('60%'));
 * ```
 *
 * @param {Handlers<T>} handlers The handlers to invoke. Can be empty, in which case the
 *   defaultHandler will be invoked. A fraction of 0 disables the handler, where it will not be
 *   called.
 * @param {HandlerFunction<T>} defaultHandler The default handler to invoke, if none of the handlers
 *   were invoked.
 * @return {T} The result of calling the randomly selected partition.
 * @throws {TypeError} If the no handler is enabled, and no default handler is provided.
 */
function partitionFraction<T>(
  handlers: Handlers<T>,
  defaultHandler: ?HandlerFunction<T> = null
): T {
  let key = Math.random(),
    lastValid = null;

  for (const [fraction, handler] of handlers) {
    // Disabled.
    if (!fraction) continue;

    if (key < fraction) {
      return handler();
    } else {
      key -= fraction;
      lastValid = handler;
    }
  }

  if (defaultHandler) {
    return defaultHandler();
  }

  // Bad arithmetic :)
  if (lastValid) {
    return lastValid();
  }

  throw new TypeError('received no enabled handlers');
}

module.exports = partitionFraction;


src/index.ts
type HandlerFunction<T> = () => T;
type Handler<T> = [number, HandlerFunction<T>];
type Handlers<T> = Array<Handler<T>>;

/**
 * Randomly invokes a handler based on the provided fractional call rates. Has a much cleaner
 * recursive implementation in a purely functional language where we'd recurse through a linked list
 * ;).
 *
 * Example:
 *
 * ```js
 * // Randomly invoke one of following arrow functions based on the given fraction.
 * // e.g. it'll invoke the first function ~10% of the time if called repeatedly.
 * invokeFraction([
 *   [0.1, () => console.log('10%')],
 *   [0.3, () => console.log('30%')],
 *   // The number here doesn't matter, provided it's >0, as it's the last partition.
 *   [0.6, () => console.log('60%')],
 * ]);
 * ```
 *
 * The `invokeFraction` function also supports a `defaultHandler` parameter, which stands in for the
 * last item of the handlers array. Note that this is the recommended way to avoid letting the last
 * item in the handlers array be invoked for the remaining partition, rather than the defined
 * fraction.
 *
 * ```js
 * invokeFraction([
 *   [0.1, () => console.log('10%')],
 *   [0.3, () => console.log('30%')],
 * ], () => console.log('60%'));
 * ```
 *
 * @param {Handlers<T>} handlers The handlers to invoke. Can be empty, in which case the
 *   defaultHandler will be invoked. A fraction of 0 disables the handler, where it will not be
 *   called.
 * @param {HandlerFunction<T>} defaultHandler The default handler to invoke, if none of the handlers
 *   were invoked.
 * @return {T} The result of calling the randomly selected partition.
 * @throws {TypeError} If the no handler is enabled, and no default handler is provided.
 */
function partitionFraction<T>(
  handlers: Handlers<T>,
  defaultHandler: HandlerFunction<T> | null | undefined = null
): T {
  let key = Math.random(),
    lastValid = null;

  for (const [fraction, handler] of handlers) {
    // Disabled.
    if (!fraction) continue;

    if (key < fraction) {
      return handler();
    } else {
      key -= fraction;
      lastValid = handler;
    }
  }

  if (defaultHandler) {
    return defaultHandler();
  }

  // Bad arithmetic :)
  if (lastValid) {
    return lastValid();
  }

  throw new TypeError('received no enabled handlers');
}

export default partitionFraction;
